File: client/lib/api/v2/schemas/reports.ts
/**
 * Reports Bundle Schemas
 */

import { z } from "zod";
import { ScenarioMappingSchema, SimulationMappingSchema } from "./base";

export const ProfileMetricsSchema = z.object({
  averageScore: z.any(), // Complex nested structure
  completionPercentage: z.any(),
  firstAttemptPassRate: z.any(),
  highestScore: z.any(),
  messagesPerSession: z.any(),
  personaResponseTimes: z.any(),
  sessionEfficiency: z.any(),
  stagnationRate: z.any(),
  timeSpent: z.any(),
  totalAttempts: z.any(),
});

export const ProfileDataEnhancedSchema = z.object({
  profileId: z.string(),
  firstName: z.string(),
  lastName: z.string(),
  alias: z.string(),
  role: z.string(),
  metrics: ProfileMetricsSchema,
});

export const ReportsBundleResponseSchema = z.object({
  data: z.array(ProfileDataEnhancedSchema),
  scenario_mapping: ScenarioMappingSchema,
  simulation_mapping: SimulationMappingSchema,
});

export type ProfileDataEnhanced = z.infer<typeof ProfileDataEnhancedSchema>;
export type ProfileMetrics = z.infer<typeof ProfileMetricsSchema>;
export type ReportsBundleResponse = z.infer<typeof ReportsBundleResponseSchema>;

File: server/app/schemas/reports.py
"""Reports bundle schemas."""


from pydantic import BaseModel

from .analytics import MetricResponse
from .base import ScenarioMapping, SimulationMapping


class ProfileMetrics(BaseModel):
    """Profile metrics - each metric is a full MetricResponse object."""

    averageScore: MetricResponse
    completionPercentage: MetricResponse
    firstAttemptPassRate: MetricResponse
    highestScore: MetricResponse
    messagesPerSession: MetricResponse
    personaResponseTimes: MetricResponse
    sessionEfficiency: MetricResponse
    stagnationRate: MetricResponse
    timeSpent: MetricResponse
    totalAttempts: MetricResponse


class ProfileDataEnhanced(BaseModel):
    """Enhanced profile data row."""

    profileId: str
    firstName: str
    lastName: str
    alias: str | None
    role: str
    metrics: ProfileMetrics


class ReportsBundleResponse(BaseModel):
    """Reports bundle response with entity mappings."""

    data: list[ProfileDataEnhanced]
    scenario_mapping: ScenarioMapping
    simulation_mapping: SimulationMapping
